design/storage-class-device-set.md

```diff
@@ -0,0 +1,487 @@
+# Rook StorageClassDeviceSets
+
+**Target version**: Rook 1.1
+
+## Background
+
+The primary motivation for this feature is to take advantage of the mobility of
+storage in cloud-based environments by defining storage based on sets of devices
+consumed as block-mode PVCs. In environments like AWS you can request remote
+storage that can be dynamically provisioned and attached to any node in a given
+Availability Zone (AZ). The design can also accommodate non-cloud environments
+through the use of local PVs.
+
+## StorageClassDeviceSet struct
+
+```go
+// File: pkg/apis/ceph.rook.io/v1/types.go
+package v1
+
+import (
+    v1 "k8s.io/api/core/v1"
+
+    rook "github.com/rook/rook/pkg/apis/rook.io/v1alpha2"
+)
+
+type StorageClassDeviceSets []StorageClassDeviceSet
+
+struct StorageClassDeviceSet {
+    Name                 string // A unique identifier for the set
+    Count                int // Number of devices in this set
+    Resources            v1.ResourceRequirements // Requests/limits for the devices
+    Placement            rook.Placement // Placement constraints for the devices
+    Config               map[string]string // Provider-specific device configuration
+    volumeClaimTemplates []v1.PersistentVolumeClaimSpec // List of PVC templates for the underlying storage devices
+}
+```
+
+A provider will be able to use the `StorageClassDeviceSet` struct to describe
+the properties of a particular set of `StorageClassDevices`. In this design, the
+notion of a "`StorageClassDevice`" is an abstract concept, separate from
+underlying storage devices. There are three main aspects to this abstraction:
+
+1. A `StorageClassDevice` is both storage and the software required to make the
+storage available and manage it in the cluster. As such, the struct takes into
+account the resources required to run the associated software, if any.
+1. A single `StorageClassDevice` could be comprised of multiple underlying
+storage devices, specified by having more than one item in the
+`volumeClaimTemplates` field.
+1. Since any storage devices that are part of a `StorageClassDevice` will be
+represented by block-mode PVCs, they will need to be associated with a Pod so
+that they can be attached to cluster nodes.
+
+A `StorageClassDeviceSet` will have the following fields associated with it:
+
+* **name**: A name for the set. **[required]**
+* **count**: The number of devices in the set. **[required]**
+* **resources**: The CPU and RAM requests/limits for the devices. Default is no
```

Comment:
I may have missed where precedence is stated in the doc. Is it stated the override precedence order of the device sets compared to other parts of the PR?

Reply:
Agreed we need to be clear about the precedence. I would expect that the `resources` defined here are highest precedence, with the `resources` at the top level of the CephCluster CR being lower precedence.